presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPageSink.java

```diff
@@ -84,122 +86,72 @@ public CassandraRecordSink(
             insert.value(columnName, bindMarker());
         }
         this.insert = cassandraSession.prepare(insert);
-
-        values = new ArrayList<>(columnTypes.size() + 1);
     }
 
     @Override
-    public void beginRecord()
+    public CompletableFuture<?> appendPage(Page page)
     {
-        checkState(field == -1, "already in record");
-
-        field = 0;
-        values.clear();
-        if (generateUUID) {
-            values.add(UUID.randomUUID());
-        }
-    }
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            List<Object> values = new ArrayList<>(columnTypes.size() + 1);
+            if (generateUUID) {
+                values.add(UUID.randomUUID());
+            }
 
-    @Override
-    public void finishRecord()
-    {
-        checkState(field != -1, "not in record");
-        checkState(field == columnTypes.size(), "not all fields set");
-        field = -1;
-        cassandraSession.execute(insert.bind(values.toArray()));
-    }
-
-    @Override
-    public void appendNull()
-    {
-        append(null);
-    }
+            for (int channel = 0; channel < page.getChannelCount(); channel++) {
+                appendColumn(values, page, position, channel);
+            }
 
-    @Override
-    public void appendBoolean(boolean value)
-    {
-        append(value);
+            cassandraSession.execute(insert.bind(values.toArray()));
+        }
+        return NOT_BLOCKED;
     }
 
-    @Override
-    public void appendLong(long value)
+    private void appendColumn(List<Object> values, Page page, int position, int channel)
     {
-        Type columnType = columnTypes.get(field);
-        if (DATE.equals(columnType)) {
-            append(DATE_FORMATTER.print(TimeUnit.DAYS.toMillis(value)));
+        Block block = page.getBlock(channel);
+        Type type = columnTypes.get(channel);
+        if (block.isNull(position)) {
+            values.add(null);
         }
-        else if (INTEGER.equals(columnType)) {
-            append(((Number) value).intValue());
+        else if (BOOLEAN.equals(type)) {
+            values.add(type.getBoolean(block, position));
         }
-        else if (REAL.equals(columnType)) {
-            append(intBitsToFloat((int) value));
+        else if (BIGINT.equals(type)) {
```

Comment:
I wonder if we would benefit from introducing a `TypeVisitor`, which would help controlling completeness (or conscious incompleteness) of  `if`-cascades like this.

Reply:
Having a type-safe visitor in the SPI for all the standard types seems useful. I wonder if @martint agrees.

Reply:
i captured the idea as #9410